Implement CPU instruction benchmarks

aochagavia opened the pull request:
This PR is a first step towards closing #1385. It implements a basic benchmarking setup suitable for CI, but doesn't hook it up yet to GitHub actions. The code is contained in its own crate (`ci-bench`).

Relevant features (check out the [readme](https://github.com/aochagavia/rustls/blob/ci-bench/ci-bench/README.md) for details):

1. Count CPU instructions for ~20 scenarios.
2. Separately measure server and client side CPU instructions.
3. Compare results between a baseline and a candidate, and generate a user-friendly overview in markdown (see the first comment to this PR as an example).

Non-features:

1. Aggregate statistics to better identify where a change in CPU counts comes from ([suggested](https://github.com/rustls/rustls/issues/1385#issuecomment-1678684307) by @djc). I'd rather get this out of the door in its current form, so we have something to iterate upon, and tweak it later if necessary.
2. GitHub Actions integration. I'd rather postpone this to focus on the benchmarking setup itself.
4. Benchmarks for memory usage. I still need to think about a reasonable setup for this one.

### Open question: significance threshold

Which instruction count diffs should be marked as significant (i.e. requiring attention of a reviewer)? I'm leaning towards a threshold of 0.2% because __it is the lowest change that can be confirmed or disproved by time-based benchmarks__.

If the instruction counts for a PR are suspect, we will want to confirm there is truly a performance regression through time-based benchmarks (remember that the change in executed CPU instructions is not necessarily in the same proportion as the change in wall time). Assuming a worst-case scenario where the wall time is affected by 5x the change in CPU instructions, a 0.2% change in instructions would amount to a 1% change in runtime. Time-based benchmarks take around 800 µs on my machine, and a difference of 1% (~8 µs) is just high enough to be detected through paired benchmarking.

There is a chance that the threshold is too low, and might result in noise (i.e. changes in instruction counts that cannot be reliably confirmed by time-based benchmarks). Still, as @djc [suggested](https://github.com/rustls/rustls/issues/1385#issuecomment-1671549976), it makes sense to start with a low threshold in order to gain experience. We can always tweak it later if it gets too annoying.

### Sidenote on the benchmarking code

The benchmarks themselves were adapted from [here](https://github.com/rustls/rustls/blob/6bdaf04e7af97821485da115b5436913fedea874/rustls/examples/internal/bench.rs). It would be nice if you can double check they are correct, because I'm not familiar enough with rustls to do so with confidence.

aochagavia commented:
Below you can see the generated markdown output after comparing two benchmark runs. The baseline numbers are from current main, the candidate numbers are from a second run, with fake changes to simulate a performance regression in data transfer for TLS 1.2

---

# Benchmark results
### Noteworthy instruction count differences
| Scenario | Baseline | Candidate | Diff 
| --- | --- | --- | --- |
| transfer_no_resume_1.2_rsa_aes_client | 7767624 | 7789541 | 21917 (0.28%) |
| transfer_no_resume_1.2_rsa_aes_server | 8775849 | 8798099 | 22250 (0.25%) |
### Other instruction count differences
<details>
<summary>Click to expand</summary>

| Scenario | Baseline | Candidate | Diff 
| --- | --- | --- | --- |
| handshake_no_resume_1.3_ecdsa_chacha_server | 1621211 | 1621487 | 276 (0.02%) |
| handshake_session_id_1.3_ecdsa_aes_server | 1646752 | 1646524 | -228 (-0.01%) |
| handshake_no_resume_1.3_ecdsa_aes_server | 1618360 | 1618562 | 202 (0.01%) |
| handshake_tickets_1.3_ecdsa_aes_server | 1644306 | 1644129 | -177 (-0.01%) |
| transfer_no_resume_1.3_rsa_aes_client | 7780252 | 7779666 | -586 (-0.01%) |
| handshake_tickets_1.3_ecdsa_chacha_server | 1648326 | 1648444 | 118 (0.01%) |
| handshake_tickets_1.2_rsa_aes_client | 4333191 | 4332957 | -234 (-0.01%) |
| handshake_session_id_1.3_ecdsa_chacha_server | 1650284 | 1650201 | -83 (-0.01%) |
| transfer_no_resume_1.3_ecdsa_aes_server | 8784596 | 8784197 | -399 (-0.00%) |
| transfer_no_resume_1.3_ecdsa_aes_client | 7781504 | 7781850 | 346 (0.00%) |
| transfer_no_resume_1.3_rsa_aes_server | 8783940 | 8784306 | 366 (0.00%) |
| handshake_session_id_1.2_rsa_aes_client | 4328918 | 4329096 | 178 (0.00%) |
| handshake_no_resume_1.2_rsa_aes_client | 4316722 | 4316886 | 164 (0.00%) |
| handshake_session_id_1.3_rsa_aes_server | 11545656 | 11546084 | 428 (0.00%) |
| handshake_no_resume_1.2_rsa_aes_server | 11328165 | 11328536 | 371 (0.00%) |
| handshake_tickets_1.3_rsa_chacha_client | 4227132 | 4227018 | -114 (-0.00%) |
| handshake_tickets_1.3_rsa_aes_server | 11545061 | 11545320 | 259 (0.00%) |
| handshake_tickets_1.3_rsa_aes_client | 4217162 | 4217254 | 92 (0.00%) |
| handshake_session_id_1.2_rsa_aes_server | 11343190 | 11342955 | -235 (-0.00%) |
| transfer_no_resume_1.3_ecdsa_chacha_client | 11245701 | 11245922 | 221 (0.00%) |
| handshake_session_id_1.3_rsa_aes_client | 4217144 | 4217225 | 81 (0.00%) |
| handshake_no_resume_1.3_rsa_chacha_server | 11527772 | 11527991 | 219 (0.00%) |
| transfer_no_resume_1.3_rsa_chacha_client | 11243310 | 11243488 | 178 (0.00%) |
| transfer_no_resume_1.3_ecdsa_chacha_server | 12218266 | 12218112 | -154 (-0.00%) |
| handshake_no_resume_1.3_ecdsa_aes_client | 14018509 | 14018358 | -151 (-0.00%) |
| handshake_session_id_1.3_rsa_chacha_server | 11556973 | 11556867 | -106 (-0.00%) |
| handshake_tickets_1.3_rsa_chacha_server | 11556880 | 11556979 | 99 (0.00%) |
| handshake_tickets_1.2_rsa_aes_server | 11348871 | 11348786 | -85 (-0.00%) |
| handshake_no_resume_1.3_rsa_aes_client | 4213207 | 4213238 | 31 (0.00%) |
| handshake_no_resume_1.3_ecdsa_chacha_client | 14021170 | 14021069 | -101 (-0.00%) |
| handshake_no_resume_1.3_rsa_aes_server | 11517858 | 11517928 | 70 (0.00%) |
| handshake_tickets_1.3_ecdsa_chacha_client | 14022476 | 14022391 | -85 (-0.00%) |
| handshake_tickets_1.3_ecdsa_aes_client | 14019868 | 14019948 | 80 (0.00%) |
| transfer_no_resume_1.3_rsa_chacha_server | 12218085 | 12218115 | 30 (0.00%) |
| handshake_session_id_1.3_ecdsa_chacha_client | 14022384 | 14022416 | 32 (0.00%) |
| handshake_session_id_1.3_rsa_chacha_client | 4227044 | 4227040 | -4 (-0.00%) |
| handshake_session_id_1.3_ecdsa_aes_client | 14019792 | 14019783 | -9 (-0.00%) |
| handshake_no_resume_1.3_rsa_chacha_client | 4223239 | 4223239 | 0 (0.00%) |
</details>

djc commented on ci-bench/Cargo.toml:
```diff
@@ -0,0 +1,22 @@
+description = "Rustls CPU instruction counting benchmarks."
+publish = false
+
+# Prevent this from interfering with workspaces
```
Why don't we want this in the workspace?

aochagavia replied:
Hmm I actually copied this from [connect-tests](https://github.com/rustls/rustls/blob/6bdaf04e7af97821485da115b5436913fedea874/connect-tests/Cargo.toml). I haven't worked much with workspaces in the past, and I assumed this was necessary for "infrastructure" code. Is it superfluous?

djc replied:
I would propose we ditch it and make this crate part of the workspace, so that it's included in CI by default.

aochagavia replied:
All right!

cpu replied:
The connect tests are a special case because they're hitting 3rd party servers over the internet and so can be prone to flakes outside of our control due to bad internet weather, cert expiry, etc.

aochagavia replied:
Hmmm... One drawback of this is that the CI now enforces that the benchmarking code compiles with the minimum supported version. Or is [this](https://github.com/rustls/rustls/actions/runs/5891298366/job/15979135819?pr=1404) an unrelated failure? It also seems to skew the coverage measurements (and I assume has an impact on the length of some CI checks).

What is the benefit of having the new crate be part of the workspace? We are going to add a dedicated CI job for it anyway, so right now I can only see the drawbacks.

djc replied:
That failure looks unrelated. And you actually set `rust-version` in the Cargo metadata for this new crate, is that a lie?

aochagavia replied:
Hmm I should have looked more closely when I copy-pasted the original `Cargo.toml` file. My natural choice would be to have no minimum rust version for this kind of code, assuming that developers of the library will always be using an up-to-date compiler. Is that ok? Or are there reasons to have a minimum rust version even for code that is not part of the library?

djc replied:
Actually the MSRV workflow in CI is already targeted at `-p rustls` only, so there should be no problem avoiding specifying a `rust-version` here and just using the stable compiler as used for most of the CI jobs.

aochagavia replied:
All right, I have removed the `rust-version` line from `Cargo.toml`

djc commented on ci-bench/src/util/mod.rs:
```diff
@@ -0,0 +1,12 @@
+pub mod config;
+pub mod key;
+
+pub fn wait_till_stdin_closed(reader: &mut dyn Read, buf: &mut [u8]) {
```
Given that all its callers are in `main.rs`, I would just move this there.

djc commented on ci-bench/src/util/config.rs:
```diff
@@ -0,0 +1,158 @@
+    clientauth: ClientAuth,
+    resume: ResumptionParam,
+) -> ClientConnection {
+    let client_config = Arc::new(make_client_config(params, clientauth, resume));
```
I would prefer inlining `make_client_config()` and `make_server_config()` into these functions.

djc commented on ci-bench/src/rustls_facade.rs:
```diff
@@ -0,0 +1,60 @@
+///
+/// Prepends a header consisting of the length of the TLS data (encoded as a big-endian u32), so the
+/// other side can know how many bytes to expect.
+pub fn send<T: SideData>(
```
Can we reuse library elements like `complete_io()` instead?

aochagavia replied:
Thanks for the tip, I didn't know about `complete_io`. I have pushed changes to start using it.

djc commented on ci-bench/src/main.rs:
```diff
@@ -0,0 +1,267 @@
+) -> anyhow::Result<()> {
+    handshake_client_side(client, reader, writer, false)?;
+
+    let mut recv_buf = Vec::new();
```
Clippy should flag this, should be `vec![0, plaintext_size as usize * 2]`.

djc commented on ci-bench/src/main.rs:
```diff
@@ -0,0 +1,267 @@
+        benchmarks.push(
+            Benchmark::new(
+                format!("new_conn_{}_{tag}", resumption_param.label()),
+                move |_reader, _writer| {
```
The `SideFn` stuff seems pretty hard to understand to me. Could we unroll this into an actual trait with two implementing types?

aochagavia replied:
I'm not sure how a trait would fit here, could you expand on the aspects that are currently hard to understand?

For context, each side of a benchmark (client and server) runs in its own subprocess. To make this possible, each benchmark must specify how to run each side. A `SideFn` is a function that runs one side of the benchmark: it needs access to a reader and a writer, and it returns a result.

djc replied:
In practice, the `SideFn` abstraction seems to be abstracting over (a) running `make_client_config()` vs `make_server_config()` and (b) `handshake_client_side()` vs `handshake_server_side()`. If that's the goal of the abstraction, I think it would be clearer to have a trait that has these two required methods. There's also a bunch of duplication between the two `SideFn` implementations which would be nice to reduce if possible.

aochagavia replied:
I see. I got a bit carried away by the desire to keep the benchmarking stuff generic, but now I realize it makes the code more difficult to read. I have replaced the usage of closures with an enum to specify which code should be run for the benchmark (i.e. only the handshake vs. handshake + transfer). Haven't pushed yet but will give you a head's up when I do.

djc commented on ci-bench/src/bench_lib/compare.rs:
```diff
@@ -0,0 +1,181 @@
+    println!("| --- | --- | --- | --- |");
+    for diff in diffs {
+        println!(
+            "| {} | {} | {} | {} ({:.2}%) |",
```
Can we do right-aligned numbers in Markdown? What about table cell background colors (green for significantly faster, red for significantly slower)?

aochagavia replied:
Aligning is possible, but colors aren't. The Rust CI uses emoji to work around this. We could do something like follows. What do you think?

### Regression

| Scenario | Baseline | Candidate | Diff 
| --- | ---: | ---: | ---: |
| transfer_no_resume_1.2_rsa_aes_client | 7767624 | 7789541 | ⚠️ 21917 (0.28%) |
| transfer_no_resume_1.2_rsa_aes_server | 8775849 | 8798099 | ⚠️ 22250 (0.25%) |

### Improvement

| Scenario | Baseline | Candidate | Diff 
| --- | ---: | ---: | ---: |
| transfer_no_resume_1.2_rsa_aes_client | 7767624 | 7789541 | ✅ -21917 (-0.28%) |
| transfer_no_resume_1.2_rsa_aes_server | 8775849 | 8798099 | ✅ -22250 (-0.25%) |

djc replied:
Emoji seems good!

djc commented on ci-bench/src/bench_lib/runner.rs:
```diff
@@ -0,0 +1,90 @@
+use std::io;
+
+/// Run a single bench at the specified side
+pub fn run_single(
```
It seems strange for this to take a `&BenchmarkCollection` + `bench_index` instead of just a single `&Benchmark`.

djc commented on ci-bench/src/bench_lib/benchmark.rs:
```diff
@@ -0,0 +1,158 @@
+use std::io::{Read, Write};
+
+/// A collection of benchmark scenarios
+pub struct BenchmarkCollection {
```
What is the value of this type? It seems we could just pass a `Vec<Benchmark>` around which would be more straightforward.

aochagavia replied:
The type is meant to encode the invariant that the collection of benchmarks has been successfully validated. I was hoping it would make the code easier to read, but I can remove it if people have a clear preference for a plain `Vec`.

djc replied:
Currently I think we have only one place that create the `Vec<Benchmark>`, so for now it feels like overkill to add a wrapper for this -- instead we can just do the validation in that function.

aochagavia replied:
Done 😉

djc commented on ci-bench/src/main.rs:
```diff
@@ -0,0 +1,267 @@
+}
+
+/// Returns the benchmark params to use for each group of benchmarks
+fn all_benchmark_params() -> Vec<BenchmarkParam> {
```
Looks like this could be a `const` slice `&[BenchmarkParam]` instead, and we could ditch the trivial `BenchmarkParam::new()` method?

aochagavia replied:
Great idea! `const` doesn't work here, but `static ALL_BENCHMARK_PARAMS: &'static [BenchmarkParam]` does

djc commented on ci-bench/src/bench_lib/benchmark.rs:
```diff
@@ -0,0 +1,158 @@
+    }
+
+    /// Returns an error if the benchmarks are invalid (see [`Self::new`])
+    fn validate_benchmarks(benchmarks: &[Benchmark]) -> anyhow::Result<()> {
```
As a single-caller function, I would just inline this into the caller.

djc reviewed (commented):
There seem to be a lot of extra abstractions that don't add much value. I would suggest simplifying this a bunch, see the specific suggestions.

cpu commented on ci-bench/src/bench_lib/cachegrind.rs:
```diff
@@ -0,0 +1,237 @@
+    }
+
+    /// Returns the CPU architecture
+    fn get_arch() -> anyhow::Result<String> {
+        let output = Command::new("uname")
+            .arg("-m")
+            .stdout(Stdio::piped())
+            .output()
+            .context("Failed to run `uname` to determine CPU architecture.")?;
+
+        let arch = String::from_utf8(output.stdout)
+            .context("`-uname -m` returned invalid unicode.")?
+            .trim()
+            .to_owned();
+
+        Ok(arch)
+    }
```
Could this be replaced with https://doc.rust-lang.org/std/env/consts/constant.ARCH.html ?

aochagavia replied:
I'm afraid this could break in unexpected ways, because `uname -m` does not always use the same string representation of an architecture as the `ARCH` constant (see [this list](https://stackoverflow.com/a/45125525/2110623) for details).

But there is an even better option! I just discovered that the `arch` argument to `setarch` is optional since the 2.33 release (in 2018), so we can remove the `get_arch` function in its entirety.

cpu replied:
> But there is an even better option! I just discovered that the arch argument to setarch is optional since the 2.33 release (in 2018), so we can remove the get_arch function in its entirety.

Ah perfect, that sounds even better. Nice catch.

cpu commented on ci-bench/src/bench_lib/cachegrind.rs:
```diff
@@ -0,0 +1,237 @@
+            .arg("-R")
+            .arg("valgrind")
+            .arg("--tool=cachegrind")
+            .arg("--cache-sim=no")
```
Maybe worth a comment similar to the `setarch` comment that mentions this disables collecting cache miss data? I had to read the fine manual to figure out what this arg does.

It also looks like this defaults to `no` (similar to `--branch-sim`). Is `cache-sim` important enough to be explicitly disabled, and if so, should `branch-sim` get the same treatment?

aochagavia replied:
Cache simulation is only disabled by default since the 3.21 release, which is pretty recent (e.g. Debian 12 ships valgrind 3.19). Branch prediction stats seem to be disabled by default already in 3.19, so probably don't need to be explicitly disabled. I'll add comments explaining the relevant arguments to valgrind.

cpu replied:
> Cache simulation is only disabled by default since the 3.21 release, which is pretty recent (e.g. Debian 12 ships valgrind 3.19)

Ahh! good call out. Thanks. I was looking at the upstream docs without any sense of what's shipped.

>  I'll add comments explaining the relevant arguments to valgrind. Do you think that would be enough here?

Yes, that sounds great. Thank you.

cpu commented on ci-bench/src/bench_lib/cachegrind.rs:
```diff
@@ -0,0 +1,237 @@
+        }
+
+        let instruction_count = parse_cachegrind_output(&self.output_file)?;
+        std::fs::remove_file(&self.output_file).ok();
```
would there be any value in an option to leave this output file in place for manual inspection?

aochagavia replied:
I think that would only be necessary for local debugging if things break in weird ways. That sounds improbable, so I'd rather have the developer comment out the line of code than adding an option to the CLI. Or do you envision another use case than debugging?

cpu replied:
Nope, that makes sense to me. Just spitballing :-)

cpu commented on ci-bench/src/bench_lib/compare.rs:
```diff
@@ -0,0 +1,188 @@
+}
+
+/// Compares all benchmark results between a baseline and a candidate, printing an overview to
+/// stdout
```
tiny nit: missing trailing `.` :-)
```suggestion
/// stdout.
```

cpu commented on ci-bench/src/bench_lib/mod.rs:
```diff
@@ -0,0 +1,17 @@
+pub use compare::compare;
+pub use runner::{run_all, run_single};
+
+pub fn black_box<T>(dummy: T) -> T {
```
What's the story with this bit? Could you add a comment motivating this fn? 

I have passing familiarity with `core::hint::black_box` and wonder why something custom is required here.

aochagavia replied:
I didn't know `core::hint::black_box` was stabilized last December! I'd gladly use it instead of a custom implementation if we aren't required to support a lower Rust version in the benchmarking code (currently being discussed [here](https://github.com/rustls/rustls/pull/1404#discussion_r1297432171)).

aochagavia replied:
Update: I've updated the code to use the `black_box` from the standard library

cpu replied:
Awesome, thanks!

cpu commented on ci-bench/src/rustls_facade.rs:
```diff
@@ -0,0 +1,60 @@
+use rustls::{ConnectionCommon, SideData};
+use std::io::{Read, Write};
+
+/// Sends data to the other side of the connection through the provided writer
```
editorial nit: missing `.`
```suggestion
/// Sends data to the other side of the connection through the provided writer.
```

cpu reviewed (commented):
This is really cool!

aochagavia commented on ci-bench/src/bench_lib/benchmark.rs:
```diff
@@ -0,0 +1,134 @@
+
+    /// Configure this benchmark to subtract the instruction count of the referenced benchmark when
+    /// reporting results
+    pub fn exclude_setup_instructions(mut self, name: String) -> Self {
```
I'll go with `String` everywhere 👍

aochagavia commented on ci-bench/README.md:
```diff
@@ -0,0 +1,110 @@
+# CI Bench
+
+This crate is meant for CI benchmarking. It measures CPU instructions using cachegrind, outputs
```
Will do!

aochagavia commented on ci-bench/src/bench_lib/cachegrind.rs:
```diff
@@ -0,0 +1,237 @@
+        })
+    }
+
+    /// Runs the benchmark at the specified index and returns the instruction counts for each side
+    pub fn run_bench(
```
Did you end up discovering? The index is necessary because it is passed to the child process (so it knows which specific benchmark it should run). Please let me know if you have any ideas to make this clearer 😉

cpu replied:
I did thanks, sorry I should have self-resolved this at the end of my review.

aochagavia commented:
Thanks for the reviews! I've resolved everything that could be resolved and left the rest open so we can discuss.

@ctz Next week I'll have a look at the issue with the resumed handshakes, thanks for pointing it out

aochagavia commented:
I just pushed changes to fix the resumed handshake benchmarks. See below a bunch of instruction counts as an example. Do they look sensible to you now, @ctz?

| Scenario | Count |
| --- | ---: |
| handshake_no_resume_1.2_rsa_aes_client | 4102933 |
| handshake_session_id_1.2_rsa_aes_client | 164924 |
| handshake_tickets_1.2_rsa_aes_client | 181010 |
| handshake_no_resume_1.2_rsa_aes_server | 11603329 |
| handshake_session_id_1.2_rsa_aes_server | 173655 |
| handshake_tickets_1.2_rsa_aes_server | 195085 |
| handshake_no_resume_1.3_rsa_aes_client | 4195395 |
| handshake_session_id_1.3_rsa_aes_client | 1148836 |
| handshake_tickets_1.3_rsa_aes_client | 1150432 |
| handshake_no_resume_1.3_rsa_aes_server | 11797230 |
| handshake_session_id_1.3_rsa_aes_server | 1222456 |
| handshake_tickets_1.3_rsa_aes_server | 1227708 |

There is, however, something weird going on with the instruction counts for the `transfer` scenario. Suddenly, they have become noisy to the point of being useless: there are false positives above 2% on every run, even without recompiling the binary! It happens consistently, but only on the client side.

Cachegrind has a `cg_diff` command to compare the instruction counts between two runs, and the variation seems to be caused by `memset` and `memcpy` calls. My research suggests there is something going on with internal buffering inside the rustls client (i.e. in some runs, the buffers grow very big, while in others they remain small, causing a big difference in `memset` and `memcpy` calls). I'll be digging into it, also by looking at previous versions of the benchmarks that generate results without the noise. If anyone has a hunch about why this is happening, please share!

aochagavia commented:
Good news, I have confirmed the instruction count noise is due to buffering, and have a local fix that works for the transfer benchmark. Now I'll see if I can also bring down the noise in the handshakes, and after some polishing the PR should be ready for a final review :)

ctz commented:
> Do they look sensible to you now, @ctz?

Yes! That looks a lot better.

aochagavia commented:
Just pushed a new version which addresses the remaining feedback from @ctz and @djc, as well as solving the noise problems. Would you mind to have another look?

In the meantime I'll be looking into the necessary tweaks to make the CI go green again ;)

aochagavia commented on ci-bench/README.md:
```diff
@@ -0,0 +1,110 @@
+# CI Bench
```
Oops, thanks!

djc commented on ci-bench/src/cli.rs:
```diff
@@ -0,0 +1,38 @@
+}
+
+#[derive(Copy, Clone, ValueEnum)]
+pub enum Side {
```
Maybe we can use `rustls::Side` instead?

aochagavia replied:
Unfortunately we need our own type, so we can derive `ValueEnum` which is required by `clap`

djc replied:
I guess. Alternatively we could have rustls implement `FromStr` for `Side`, but that might not be much better.

aochagavia replied:
I can't imagine that `FromStr` implementation would be useful anywhere else, so in my eyes the current solution is preferrable

djc commented on ci-bench/src/bench_lib/benchmark.rs:
```diff
@@ -0,0 +1,179 @@
+use crate::cli::Side;
+use crate::util::key::KeyType;
```
Style nit: would be nice to split these import blocks with newlines, between `crate`, external crates and `std` imports.

djc commented on ci-bench/src/main.rs:
```diff
@@ -0,0 +1,309 @@
+    let cli = Cli::parse();
+    match cli.command {
+        Command::RunAll => {
+            let executable = std::env::args().next().unwrap();
```
Is there some way to get the executable name as part of clap? The direct use of `std::env::args()` in the context of a clap parser feels bit odd.

aochagavia replied:
I agree it's odd, but clap doesn't seem to offer this through their API (the docs have no relevant references to the word `binary` or `executable` and googling did not reveal anything either)

djc commented on ci-bench/src/main.rs:
```diff
@@ -0,0 +1,309 @@
+}
+
+/// Returns all benchmarks
+fn get_benchmarks() -> anyhow::Result<Vec<Benchmark>> {
```
Naming nit: per the API guidelines, I would avoid the `get_()` prefix here.

aochagavia replied:
Will do

djc commented on ci-bench/src/main.rs:
```diff
@@ -0,0 +1,309 @@
+}
+
+/// Runs a benchmark for the client-side of the connection
+fn run_bench_for_client(
```
I think I still maintain that this would be better if you had a `Handshake` trait or similar that takes the `new_client()`, `handshake_client_side()` (maybe these can be combined?) and `handshake_client()` side out, so there's no need to deduplicate `run_bench_for_client()` and `run_bench_for_server()`.

aochagavia replied:
I'll have a look into it

aochagavia replied:
Just pushed a commit which maces `run_bench` generic based on a `BenchStepper` trait. Does that look like what you had in mind?

djc replied:
It does seem like an improvement. I'd probably factor out the 3 common fields of the `Stepper` types into a separate type and also attach the `make_new_config()` types as methods to the `{Client,Server}SideStepper` types.

djc reviewed (commented):
I still have a few nits.

At the top level, for my taste this has too many small files that I have to keep skipping around between (but I know this is somewhat opinionated). I would probably concatenate all the `util` modules into one.

It also feels weird to me that there's a `_lib` module inside the crate. I would restructure this to have the `main.rs` contain the executable code, move the `cli` contents in that same module, and then move all of the code in `bench_lib` one level up so it can be a proper library crate (with a `lib.rs` module).

aochagavia commented:
@djc regarding your suggestions around the structure of the code, I have one general consideration and a few specific remarks.

The general consideration is that I personally prefer grouping code in a separate file when there is strong conceptual cohesion, even if that results in small files (it makes it easier for me to navigate the code, contrary to big files with less conceptual cohesion). I assumed this made things easier on other people, too, but I guess it's highly dependent on your workflow. I'll try to restrain myself a bit more, then, hoping I'll strike the right balance and you won't have to constantly remember me of it 😉.

On the specific suggestions:

- Re merging all `util::*` modules into a single `util` module. I agree that `util::config` and `util::key` could be merged, but `util::transport` has important module-level comments (those starting with `//!`) that we don't want to lose. What would you suggest we do with them? One alternative would be to leave the file alone and move it to the top level (`crate::transport`), but maybe that's just because I'm inclined to keep the file. Another alternative would be to have an inline module inside `util` (i.e. `mod transport { ... }`). Otherwise we could move the code inside `util` without ceremony, turning the module comments into normal ones `//` and ensuring they precede the relevant code (hoping future readers will go top-to-bottom and find them so they have relevant context). Any preference?
- Re merging `cli` into main. That sounds totally sensible.
- Re `bench_lib`. I agree that it is a weird choice, and while it might have made sense at the beginning (when I thought part of the code would be extracted as a library), it clearly doesn't make sense now. For that reason, instead of extracting that as a library crate, I think we should do the following (please let me know whether you agree):
  - Merge `bench_lib::runner` and `bench_lib::compare` into `main`, to get rid of the small files.
  - Move `bench_lib::cachegrind` and `bench_lib::benchmark` to the top level.

djc commented:
> * Re merging all `util::*` modules into a single `util` module. I agree that `util::config` and `util::key` could be merged, but `util::transport` has important module-level comments (those starting with `//!`) that we don't want to lose. What would you suggest we do with them? One alternative would be to leave the file alone and move it to the top level (`crate::transport`), but maybe that's just because I'm inclined to keep the file. Another alternative would be to have an inline module inside `util` (i.e. `mod transport { ... }`). Otherwise we could move the code inside `util` without ceremony, turning the module comments into normal ones `//` and ensuring they precede the relevant code (hoping future readers will go top-to-bottom and find them so they have relevant context). Any preference?

Inline modules seem like a decent middle ground here.

> * Re `bench_lib`. I agree that it is a weird choice, and while it might have made sense at the beginning (when I thought part of the code would be extracted as a library), it clearly doesn't make sense now. For that reason, instead of extracting that as a library crate, I think we should do the following (please let me know whether you agree):
>   
>   * Merge `bench_lib::runner` and `bench_lib::compare` into `main`, to get rid of the small files.
>   * Move `bench_lib::cachegrind` and `bench_lib::benchmark` to the top level.

These sound like sensible improvements.

aochagavia commented:
All right, I think I have now addressed all your comments @djc. Thanks for taking the time to review this!

cpu commented:
Nice!
